### PR TITLE
Ledger layer refactor

### PR DESCRIPTION
**Pull request by MuncleUscles**

### CODE REVIEW

File: src/modules/ledger/RadixLedger.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+                tap({
+                    complete: () => {
+                        hasSyncedStore.next(true)
+                    }
+                }),
+            ),
+            this.atomStore.getAtomObservations(address),
+        )
+
+        // this.storeSyncStatus.set(atomObservable, hasSyncedStore)
```
Comment (laszlourszuly) [inline src/modules/ledger/RadixLedger.ts]: Is this something we want to remove?

File: src/modules/ledger/RadixLedger.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        })
+
+        return subscription
+    }
+
+
+    private monitorAtomFinality = (observation: RadixAtomObservation) => {
+        const aid = observation.atom.getAid()
+
+        if (aid.toString() in this.finalityTimeouts) {
```
Comment (laszlourszuly) [inline src/modules/ledger/RadixLedger.ts]: We seem to be using `aid.toString()` quite frequently in this context. Would it make sense to use a constant of it instead?

File: src/modules/ledger/RadixNEDBAtomStore.ts [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+                observer.complete()
+            })
+        })
+    }
+
+
+    private asyncDeserialize(atom) {
+        return new Promise<RadixAtom>((resolve, reject) => {
+            setTimeout(() => {
```
Comment (laszlourszuly) [inline src/modules/ledger/RadixNEDBAtomStore.ts:167]: Out of curiosity: Why this zero-delay timeout? Is this a commonly known magic trick, or would a describing comment be justifiable here?

File: src/modules/universe/RadixBootstrapConfig.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { RadixUniverseConfig, RadixNodeDiscovery } from '../..';
+
+export interface RadixBootstrapConfig {
+    universeConfig: RadixUniverseConfig,
+    nodeDiscovery: RadixNodeDiscovery,
+    finalityTime: number,
```
Comment (laszlourszuly) [inline src/modules/universe/RadixBootstrapConfig.ts:6]: What's this `finalityTime`? It seems nice if it is what I think it is.

File: src/modules/universe/RadixNodeConnection.ts [hunk 5]
```diff
@@ -322,87 +321,28 @@ The atom may or may not have been accepted by the node.
...
-                subject.complete()
-                break
-            case 'COLLISION':
-            case 'ILLEGAL_STATE':
-            case 'UNSUITABLE_PEER':
-            case 'VALIDATION_ERROR':
-                subject.error(value + ': ' + message)
-                break
-        }
+        // subject.next({status: RadixAtomNodeStatus[value], data: notification.data})
```
Comment (laszlourszuly) [inline src/modules/universe/RadixNodeConnection.ts]: Is this something we want to remove?

File: test/integration/MetaDataAtom.spec.ts [hunk 6]
```diff
@@ -50,15 +51,12 @@ describe('RLAU-572: MetaData in Atoms', () => {
...
-        await identity1.account.openNodeConnection()
-        await identity2.account.openNodeConnection()
+        
         nodeConnection = await radixUniverse.getNodeConnection(identity1.address.getShard())
     })
 
     after(async () => {
-        await identity1.account.closeNodeConnection()
-        await identity2.account.closeNodeConnection()
+        //
```
Comment (laszlourszuly) [inline test/integration/MetaDataAtom.spec.ts]: Is this some "prevent-auto-format-to-do-what-it-is-configured-to-do" magic?

File: src/modules/account/RadixAccount.ts [hunk 7]
```diff
@@ -145,55 +155,27 @@ export default class RadixAccount {
...
 
-    /**
-     * Unsubscribes the node connection to the stream of past and future atoms associated with this address account
-     * 
-     * @returns A promise with the result of the unsubscription call
-     */
-    public closeNodeConnection = async () => {
-        this.connectionStatus.next('DISCONNECTED')
+    private _onAtomReceived = async (atomObservation: RadixAtomObservation) => {
+        this.processingAtomCounter.next(this.processingAtomCounter.getValue() + 1)
```
Comment (laszlourszuly) [inline src/modules/account/RadixAccount.ts:168]: Is this thread safe? What will this `processingAtomCounter.next(...)` provide if another thread is executed before `this.processingAtomCounter.getValue()` on this thread is called?

File: src/modules/account/RadixAccount.ts [hunk 8]
```diff
@@ -60,11 +58,20 @@ export default class RadixAccount {
...
-        }        
+        }
+
+        this.atomObservable = radixUniverse.ledger.getAtomObservations(address)
+        
+        this.atomObservable.subscribe({
+            next: this._onAtomReceived,
+        })
+
+        // radixUniverse.ledger.onSynced(this.atomObservable)
```
Comment (laszlourszuly) [inline src/modules/account/RadixAccount.ts]: Is this something we want to remove?

**Review by laszlourszuly — COMMENTED** [review]
Looks good to me! I only had some minor educational questions, nothing blocking at all. And I really like the documentation effort you've put in as well.

File: src/modules/ledger/RadixNEDBAtomStore.ts (hunk 3 above)
Comment (MuncleUscles) [reply src/modules/ledger/RadixNEDBAtomStore.ts:167]: This is a well known trick, to understand how it works you need to know about the JS event loop https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop

TL;DR JS is single threaded, using a 0 timeout will schedule the function to be executed during the next "available opportunity", unblocking the current thread so that UI can render smoothly etc.

In this case, I don't think it can be quickly describe in a comment, and it's used in multiple places, I think the function name itself captures the essence of what's happening

File: src/modules/universe/RadixBootstrapConfig.ts (hunk 4 above)
Comment (MuncleUscles) [reply src/modules/universe/RadixBootstrapConfig.ts:6]: Finality time is the delay, from receiving the last atom status update, after which the atom status can be considered final - it kind of depends on the network size, but 5s is a reasonable assumption for now.

File: test/integration/MetaDataAtom.spec.ts (hunk 6 above)
Comment (MuncleUscles) [reply test/integration/MetaDataAtom.spec.ts]: It's just me being lazy, I should take out the whole function. But also, I refuse to auto-format :D

File: src/modules/account/RadixAccount.ts (hunk 7 above)
Comment (MuncleUscles) [reply src/modules/account/RadixAccount.ts:168]: There are no threads in JS, so yes, this is safe.

File: test/integration/MetaDataAtom.spec.ts (hunk 6 above)
Comment (laszlourszuly) [reply test/integration/MetaDataAtom.spec.ts]: A man of culture! 😆

File: src/modules/account/RadixAccount.ts (hunk 7 above)
Comment (laszlourszuly) [reply src/modules/account/RadixAccount.ts:168]: Hm... but what does `async` mean then?

Comment (laszlourszuly) [reply src/modules/account/RadixAccount.ts:168]: Never mind, I can google this...

Comment (MuncleUscles) [reply src/modules/account/RadixAccount.ts:168]: It's a real fancy way to do a callback effectively, but everything always happens in the same thread
